lib/interceptor/cache.js

```diff
@@ -0,0 +1,185 @@
+          if (typeof handler.onConnect === 'function') {
+            handler.onConnect(ac.abort)
+            signal.throwIfAborted()
+          }
+
+          if (typeof handler.onHeaders === 'function') {
+            // Add the age header
+            // https://www.rfc-editor.org/rfc/rfc9111.html#name-age
+            const age = Math.round((Date.now() - value.cachedAt) / 1000)
+
+            value.rawHeaders.push(AGE_HEADER, Buffer.from(`${age}`))
+
+            handler.onHeaders(value.statusCode, value.rawHeaders, stream.resume, value.statusMessage)
+            signal.throwIfAborted()
+          }
+
+          if (opts.method === 'HEAD') {
+            if (typeof handler.onComplete === 'function') {
+              handler.onComplete(null)
+              stream.destroy()
+            }
+          } else {
+            if (typeof handler.onData === 'function') {
+              stream.on('data', chunk => {
+                if (!handler.onData(chunk)) {
+                  stream.pause()
+                }
+              })
+            }
+
+            if (typeof handler.onComplete === 'function') {
+              stream.on('end', () => {
+                handler.onComplete(value.rawTrailers ?? [])
+              })
+            }
+          }
+        } catch (err) {
+          stream.destroy(err)
+          if (!onErrorCalled && typeof handler.onError === 'function') {
+            handler.onError(err)
+            onErrorCalled = true
+          }
+        }
+      }
+
+      /**
+       * @param {import('../../types/cache-interceptor.d.ts').default.CacheStoreReadable | undefined} stream
+       */
+      const handleStream = (stream) => {
+        if (!stream) {
+          // Request isn't cached
+          return dispatch(opts, new CacheHandler(globalOpts, opts, handler))
+        }
+
+        const { value } = stream
+
+        // Dump body on error
+        if (util.isStream(opts.body)) {
+          opts.body?.on('error', () => {}).resume()
+        }
+
+        // Check if the response is stale
+        const now = Date.now()
+        if (now >= value.staleAt) {
+          if (now >= value.deleteAt) {
+            // Safety check in case the store gave us a response that should've been
+            //  deleted already
+            dispatch(opts, new CacheHandler(globalOpts, opts, handler))
+            return
+          }
+
+          if (!opts.headers) {
+            opts.headers = {}
+          }
+
+          opts.headers['if-modified-since'] = new Date(value.cachedAt).toUTCString()
+
+          // Need to revalidate the response
+          dispatch(
+            opts,
+            new CacheRevalidationHandler(
+              () => respondWithCachedValue(stream, value),
+              new CacheHandler(globalOpts, opts, handler)
+            )
+          )
+
+          return
+        }
+
+        respondWithCachedValue(stream, value)
+      }
+
+      Promise.resolve(stream).then(handleStream).catch(err => handler.onError(err))
+
```

Comment:
why wrap it?


```suggestion
      Promise.resolve(stream).then(handleStream).catch(handler.onError)
```

Reply:
The result of `createReadStream` can be a promise

Reply:
> I would prefer for this to not be a promise.

From what I see, `stream` here is coming from `CacheStore`'s `createReadStream`. People may bring their cache store implementation, and depending on the implementation the returned value could be a promise, and IMHO undici should accept that.

Let's say, what happens if someone is using redis as the backend as their cache store and all [`ioredis`](https://www.npmjs.com/package/ioredis) API are async.